clash-cores/src/Clash/Cores/Xilinx/DcFifo/Explicit.hs

```diff
@@ -0,0 +1,247 @@
+import           Clash.Explicit.Prelude
+import           Clash.Signal.Internal       (Signal (..))
+
+
+import qualified Data.Sequence               as Seq
+import           Data.String.Interpolate     (i)
+
+-- We want constructor names to correspond 1:1 to Xilinx options it ease
+-- blackbox implementation.
+{-# HLINT ignore "Use camelCase" #-}
+
+type ResetBusy = Bool
+type Full = Bool
+type Empty = Bool
+type DataCount n = Unsigned n
+
+data ReadMode
+  = Standard_FIFO
+  | First_Word_Fall_Through
+  deriving (Show, Generic)
+
+data DcImplementation
+  = Independent_Clocks_Block_RAM
+  | Independent_Clocks_Distributed_RAM
+  | Independent_Clocks_Builtin_FIFO
+  deriving (Show, Generic)
+
+data DcConfig depth = DcConfig
+  { dcDepth          :: !depth
+  , dcImplementation :: !DcImplementation
+  , dcReadMode       :: !ReadMode
+  , dcReadDataCount  :: !Bool
+  , dcWriteDataCount :: !Bool
+  , dcOverflow       :: !Bool
+  , dcUnderflow      :: !Bool
+  }
+  deriving (Show, Generic)
+
+defConfig :: KnownNat depth => DcConfig (SNat depth)
+defConfig = DcConfig
+  { dcDepth = SNat
+  , dcImplementation = Independent_Clocks_Block_RAM
+  , dcReadMode = Standard_FIFO
+  , dcReadDataCount = True
+  , dcWriteDataCount = True
+  , dcOverflow = False
+  , dcUnderflow = False
+  }
+
+data FifoState n = FifoState
+  { hsQueue      :: Seq.Seq (BitVector n)
+  , relativeTime :: Int -- ^ In order to model a FIFO in two clock domains,
+                        -- we track relative offset in order to know which
+                        -- signals to peel off.
+  } deriving Show
+
+data XilinxFifo read write depth n =
+  XilinxFifo
+    { writeReset :: Signal write ResetBusy -- ^ @wr_rst@
+    , isFull :: Signal write Full -- ^ @full@
+    , isOverflow :: Signal write Bool -- ^ @overflow@
+    , writeCount :: Signal write (DataCount depth) -- ^ @wr_data_count@
+    , readReset :: Signal read ResetBusy -- ^ @rd_rst@
+    , isEmpty :: Signal read Empty -- ^ @empty@
+    , isUnderflow :: Signal read Bool -- ^ @underflow@
+    , readCount :: Signal read (DataCount depth) -- ^ @rd_data_count@
+    , fifoData :: Signal read (BitVector n) -- ^ @dout@
+    }
+
+-- | Xilinx FIFO, see [documentation](https://docs.xilinx.com/v/u/en-US/pg057-fifo-generator)
+-- for expected behavior. Note that the behavior does not correspond exactly to
+-- the behavior of the RTL model (which doesn't correspond exactly to hardware).
+--
+-- If 'dcReadDataCount', 'dcUnderflow', 'dcOverflow', or 'dcWriteDataCount' are disabled, the relevant signals
+-- will be 'deepErrorX'
+dcFifo ::
+  forall depth n write read .
+  ( KnownNat n
+  , KnownDomain write
+  , KnownDomain read
+
+  , KnownNat depth
+  -- Number of elements should be between [2**4, 2**17] ~ [16, 131072].
+  , 4 <= depth
+  , depth <= 17
+  ) =>
+  DcConfig (SNat depth) ->
+
+  -- | @wr_clk@
+  Clock write ->
+  -- | @rd_clk@
+  Clock read ->
+  -- | Asynchronous reset @rst@
+  Reset read ->
```

Comment:
Yeah I discussed it with Christiaan and have it figured out. I've made notes for myself but I‍ haven't had time to do a review yet.

Reply:
Okay, so Christiaan explained some stuff about resets to me and we consulted the datasheet. Whether this Clash `Reset` is synchronous or asynchronous is determined by the domain called `read` here, so the Haddock should just say something like:

```suggestion
  -- | Reset for both read and write side
  Reset read ->
```

The FIFO will work fine with either `ResetKind` (and in fact, `XilinxSystem` has `Synchronous`). However, what _will_ need to be handled is the `ResetPolarity`. This will need the equivalent of an `unsafeToHighPolarity` passed over the `Reset`.

You might want to consider moving it to the write side, though (`Reset write`), as that is the only option on UltraScale builtin FIFO which we might add later.

And in the test suite, you should keep it asserted for at least one cycle of the slow clock.

Reply:
>This will need the equivalent of an unsafeToHighPolarity passed over the Reset.

We do use `unsafeToHighPolarity` in the Haskell model in order to consume the `Reset`! So I think this is correct as-is.

Reply:
You need to do the same thing in the primitive too :).

Reply:
To elaborate a bit: the Xilinx FIFO (in HDL) will currently be connected to whatever reset is given. The Xilinx IP expects a high polarity reset (if I'm not mistaken, please check this in the data sheet), so you need to make sure to do the equivalent of `unsafeToHighPolarity` in the Blackbox.

Reply:
> I use resetGen which I believe keeps the reset asserted exactly one clock cycle.
... of the clock you do the reset on, which is currently the read clock (although I suggested to move it to the write clock).

So if the write clock is the slow clock; no, you do not keep the reset asserted for at least one cycle of the slow clock ;-).

> The Xilinx IP expects a high polarity reset (if I'm not mistaken, please check this in the data sheet)
Yes, that is the case (no need to check).

Reply:
Yes, you're right @DigitalBrains1 :)

I have pushed fixes so that the slow clock is always the one fed in to the FIFO.

Reply:
Also pushed fe716a5573cc18506a03ca6f60cd297b4a2848ee @martijnbastiaan I believe this is what we want.